compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/DecodeV1.hs

```diff
@@ -501,111 +511,128 @@ decodePrim = pure . \case
-    decodeWithArgs args $ TCon <$>  mayDecode "type_ConTycon" mbCon decodeTypeConName
+    decodeWithArgs args $ TCon <$> mayDecode "type_ConTycon" mbCon decodeTypeConName
   LF1.TypeSumPrim (LF1.Type_Prim (Proto.Enumerated (Right prim)) args) -> do
-    decodeWithArgs args $ TBuiltin <$> decodePrim prim
+    decodeWithArgs args $ TBuiltin <$> (decodeImpl $ decodePrim prim)
   LF1.TypeSumPrim (LF1.Type_Prim (Proto.Enumerated (Left idx)) _args) ->
-    Left (UnknownEnum "Prim" idx)
+    throwError (UnknownEnum "Prim" idx)
   LF1.TypeSumFun (LF1.Type_Fun params mbResult) -> do
     mkTFuns
       <$> mapM decodeType (V.toList params)
       <*> mayDecode "type_FunResult" mbResult decodeType
   LF1.TypeSumForall (LF1.Type_Forall binders mbBody) -> do
     body <- mayDecode "type_ForAllBody" mbBody decodeType
-    foldr TForall body <$> traverse decodeTypeVarWithKind (V.toList binders)
+    decodeImpl $ foldr TForall body <$> traverse decodeTypeVarWithKind (V.toList binders)
   LF1.TypeSumTuple (LF1.Type_Tuple flds) ->
     TTuple <$> mapM (decodeFieldWithType FieldName) (V.toList flds)
   where
-    decodeWithArgs :: V.Vector LF1.Type -> Decode Type -> Decode Type
+    decodeWithArgs :: V.Vector LF1.Type -> DecodeImpl Type -> DecodeImpl Type
     decodeWithArgs args fun = foldl TApp <$> fun <*> traverse decodeType args
 
 
-decodeFieldWithType :: (T.Text -> a) -> LF1.FieldWithType -> Decode (a, Type)
+decodeFieldWithType :: (T.Text -> a) -> LF1.FieldWithType -> DecodeImpl (a, Type)
 decodeFieldWithType wrapName (LF1.FieldWithType name mbType) =
   (,)
     <$> decodeName wrapName name
     <*> mayDecode "fieldWithTypeType" mbType decodeType
 
-decodeFieldWithExpr :: LF1.FieldWithExpr -> Decode (FieldName, Expr)
+decodeFieldWithExpr :: LF1.FieldWithExpr -> DecodeImpl (FieldName, Expr)
 decodeFieldWithExpr (LF1.FieldWithExpr name mbExpr) =
   (,)
     <$> decodeName FieldName name
     <*> mayDecode "fieldWithExprExpr" mbExpr decodeExpr
 
-decodeTypeConApp :: LF1.Type_Con -> Decode TypeConApp
+decodeTypeConApp :: LF1.Type_Con -> DecodeImpl TypeConApp
 decodeTypeConApp LF1.Type_Con{..} =
   TypeConApp
     <$> mayDecode "typeConAppTycon" type_ConTycon decodeTypeConName
     <*> mapM decodeType (V.toList type_ConArgs)
 
-decodeTypeConName :: LF1.TypeConName -> Decode (Qualified TypeConName)
+decodeTypeConName :: LF1.TypeConName -> DecodeImpl (Qualified TypeConName)
 decodeTypeConName LF1.TypeConName{..} = do
   (pref, mname) <- mayDecode "typeConNameModule" typeConNameModule decodeModuleRef
   con <- mayDecode "typeConNameName" typeConNameName (decodeDottedName TypeConName)
   pure $ Qualified pref mname con
 
-decodePackageRef :: LF1.PackageRef -> Decode PackageRef
+decodePackageId :: TL.Text -> PackageId
+decodePackageId = PackageId . TL.toStrict
+
+-- the invariant *does not* hold: pkid `cmp` opkid = index pkid `cmp` index opkid
+-- it is only true internally for one particular encoder implementation
+type PackageRefCtx = V.Vector PackageId
```

Comment:
I'd have probably made the PackageRefCtx more opaque and just passed around a `Int -> Maybe PackageId` function, abstracting away the Vector aspect.

Reply:
Makes sense. e0606d8f67743767bee63492fb49062729af391c